aziot/src/internal/check/checks/mod.rs

```diff
@@ -0,0 +1,98 @@
+// Copyright (c) Microsoft. All rights reserved.
+
+use super::Checker;
+
+mod prelude {
+    pub use anyhow::{anyhow, Context, Error, Result};
+    pub use serde::Serialize;
+    pub use tokio::prelude::*;
```

Comment:
Please no wildcard imports, even from preludes.

Reply:
Sure, I don't mind being more explicit about what tokio types get included here.

Reply:
Good point from @massand, it looks like the tokio folks did indeed remove the prelude in 1.0 https://github.com/tokio-rs/tokio/pull/3299

Reply:
Two problems with wildcard imports:

1. `pub use foo::*;` behaves weirdly when a `*`-imported ident conflicts with an explicitly defined one. Eg if tokio exported `Foo` and you also had `pub struct Foo`, then your crate would compile fine, but the crate that uses `this_crate::Foo` later gets an error.

1. It makes it impossible for a non-IDE user to know where a particular import is coming from.

Reply:
(I'm going to assume you meant to comment on https://github.com/Azure/iot-identity-service/pull/83#discussion_r552055404)

For point 1, I totally agree with you, hence why I don't like exporting preludes to external users.
For point 2, yeaaah, that's a more compelling argument.

That said, like I said in https://github.com/Azure/iot-identity-service/pull/83#discussion_r552030877, I'm not sure point 2 matters too much in this _particular_ case, since the prelude is a) only visible to submodules, and b) used across a set of submodules that all share a nearly identical import and code structure. IMO this private prelude actually improves the readability of the individual check implementations, since it puts more emphasis on the "meat" of the code, instead of the boilerplate corresponding to importing types like `CheckResult`' from `crate::internal::check` (or alternatively, `super::super::check`).

That said, there is an argument to be made about removing certain external crate re-exports from the prelude, as while it's fairly obvious what the `Serialize` identifier corresponds to, having `Result` correspond to `anyhow::Result` could be a bit confusing at first glance. I personally don't think so, but hey, this entire debate is about personal opinion :wink: 

While this isn't a hill I'm ready to die on, I do think we should keep this little private prelude (or at least some subset of it). What do other folks think about this, @massand and/or @gordonwang0?

Reply:
>(I'm going to assume you meant to comment on #83 (comment))

Point 1 only applies here; the other place wasn't a re-export. Point 2 applies to both.

>For point 1, I totally agree with you, hence why I don't like exporting preludes to external users.

The situation I described happens between modules too, not just between crates.

https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=cd3f3d15db79329b0fa6dc6379345588

Notice how the error is reported on the *use* of `Foo` in `m3::bar`, rather than on the `prelude` mod's re-exports that created the ambiguity in the first place. If you comment out the use of the ambiguous `Foo`, you'll see the code compiles fine, ie it's a trap waiting for the first person to start using `Foo`.

>That said, there is an argument to be made about removing certain external crate re-exports from the prelude, as while it's fairly obvious what the `Serialize` identifier corresponds to, having `Result` correspond to `anyhow::Result` could be a bit confusing at first glance.

It's not about confuson but about effort. It requires more effort to determine that `Error` is not imported from anywhere explicitly and thus is presumably coming from `prelude::*`, and then determine that `prelude` defines `Error` as a re-export of `anyhow::Error`. I'd rather it be made obvious that `Error` is `anyhow::Error` without having to jump to another file to get the answer.

Code is read more often than it is written. It helps when every place consistently imports well-known things. Eg I'm used to importing `Serialize` from `serde`, etc, so when searching for where `Serialize` is defined I automatically look for where `use serde::` ought to be in the file.

>IMO this private prelude actually improves the readability of the individual check implementations

Assuming the third-party imports are removed from the prelude, the remaining ones are all from `crate::internal::check`, plus one trait defined inline in the prelude. In the first place preludes don't define anything inline, so this ought to be moved somewhere, probably to `crate::internal::check`. That means all the imports are from `crate::internal::check`. So again, it's counter-productive to have to jump through hoops to figure out that `CheckResult` is not being imported explicitly and must thus be coming from `prelude::*`, and so on.

Reply:
Huh, you know what, that's actually a pretty convincing argument you've got there. I think I might reconsider my use of preludes in my own code moving forwards...

So yeah, sure, I'll remove it.

Reply:
Yeah, I can't say I'm too keen on your spartan use of `use`.

My personal rule of thumb is that I'll use `use` when the type that's imported is unambiguous in that given context. e.g: I would never do something like `use std::collections::hash_map::IntoIter`, since `IntoIter` is far too ambigous. That said, something like `use std::sync::Arc` is totally fair game, since most of the time, `Arc` can only ever be one thing.